### ISSUE TITLE
Resolve `-Wctad-maybe-unsupported` warnings

### ISSUE THREAD

**Issue by dtzSiFive** [llvm/circt#3943]
These show up in our CI as well as in local builds, for example:

```
 /__w/circt/circt/tools/circt-reduce/Reduction.cpp:78:21: warning: 'Optional' may not intend to support class template argument deduction [-Wctad-maybe-unsupported]
  return moduleOp ? llvm::Optional(moduleOp)
                    ^
/__w/circt/circt/llvm/llvm/include/llvm/ADT/APInt.h:33:29: note: add a deduction guide to suppress this warning
template <typename T> class Optional;
                            ^
```

**Comment by dtzSiFive** [comment]
The example above and many of the others seem to be due to using forward declarations.  Including the real header may address the warning (but defeats the purpose of only using a declaration, see our `LLVM.h` as well as same from upstream MLIR).

Haven't investigated further, suggestions/ideas welcome, but at least can track this issue here.

**Comment by dtzSiFive** [comment]
These appear to all be resolved now.
